core/modules/entity/entity.controller.inc

```diff
@@ -365,7 +365,17 @@ class DefaultEntityController implements EntityControllerInterface {
         $queried_entities_by_id = array_intersect_key($queried_entities, $passed_ids);
         if (!empty($queried_entities_by_id)) {
           foreach ($queried_entities_by_id as $id => $item) {
-            cache_set($id, $item, 'cache_entity_' . $this->entityType);
+            if (!empty($this->entityInfo['entity keys']['bundle'])) {
+              $bundle_property = $this->entityInfo['entity keys']['bundle'];
+              $bundle_name = $item->$bundle_property;
+            }
+            else {
+              $bundle_name = $this->entityType;
+            }
```

Comment:
All of the code above can be substituted with a single line:

```php
list(, , $bundle_name) = entity_extract_ids($this->entityType, $item);
```

This also takes care of entities with no bundles (returning the entity type).

Reply:
Or `$item->bundle()`.

Reply:
But Entity::bundle() can be left empty for some entity classes. Take for instance `RulesConfigEntity` . I think `entity_extract_ids` is safer.

Reply:
`bundle()` is required for Backdrop entities, as per `EntityInterface::bundle()`, and `RulesConfigEntity()` does actually have that method. https://github.com/backdrop-contrib/rules/blob/1.x-2.x/includes/rules.core.inc#L38

Reply:
What I meant is that RulesConfigEntity::bundle() returns null. My point is that there are entity classes out there that don't return a bundle (although it should), whereas you always get a bundle name with entity_extract_ids

Reply:
Further entity_extract_id() is more commonly used in Backdrop - 46 hits, against 17 hits of Entity::bundle(). Granted, Entity::bundle() was introduced more recently to core than the other. Still, I'd go for the safer/ more common way of getting this.

Reply:
Three more examples of commonly used Entity extensions that return NULL for bundle(): 
- UcOrderProduct
- UcOrder
- UcCartItem

Which further supports the use of the function vs the method

Reply:
That was pretty slick! I'll incorporate that in the next rev.

Reply:
I understand your reasoning here. Realistically though, all those cases where contrib overrides `bundle()` without returning anything are wrong, and should be fixed there in the contrib project rather than core continuing to accommodate the more outdated way of doing things. I'll create an issue in Rules.